cmd/util/ledger/migrations/multiple_contract_migration.go

```diff
@@ -0,0 +1,261 @@
+	for i, err := range e.Errors {
+		_, err := fmt.Fprintf(&b, "\nerr %d: %s", i, err)
+		if err != nil {
+			panic(err)
+		}
+	}
+	return b.String()
+}
+
+var (
+	MultipleContractsSpecialMigrations = make(map[string]func(ledger.Payload, zerolog.Logger) ([]ledger.Payload, error))
+)
+
+func MultipleContractMigration(payload []ledger.Payload) ([]ledger.Payload, error) {
+	migratedPayloads := make([]ledger.Payload, 0)
+	errors := make([]error, 0)
+	// the cache keeps a record of already migrated addresses
+	// it is completely ephemeral. If lost, it will only effect performance
+	cache := map[string]struct{}{}
+	logger := zerolog.Logger{}
+
+	for _, p := range payload {
+		results, err := migrateRegister(p, cache, logger)
+		// dont fail fast... try to collect errors so multiple errors can be addressed at once
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		migratedPayloads = append(migratedPayloads, results...)
+	}
+	if len(errors) != 0 {
+		return nil, &MultipleContractMigrationError{
+			Errors: errors,
+		}
+	}
+	return migratedPayloads, nil
+}
+
+func keyToRegisterId(key ledger.Key) (flow.RegisterID, error) {
+	if len(key.KeyParts) != 3 ||
+		key.KeyParts[0].Type != state.KeyPartOwner ||
+		key.KeyParts[1].Type != state.KeyPartController ||
+		key.KeyParts[2].Type != state.KeyPartKey {
+		return flow.RegisterID{}, fmt.Errorf("key not in expected format %s", key.String())
+	}
+
+	return flow.NewRegisterID(string(key.KeyParts[0].Value), string(key.KeyParts[1].Value), string(key.KeyParts[2].Value)), nil
+}
+
+func migrateRegister(p ledger.Payload, cache map[string]struct{}, logger zerolog.Logger) ([]ledger.Payload, error) {
+	registerId, err := keyToRegisterId(p.Key)
+	if err != nil {
+		return nil, err
+	}
+	// have we migrated this?
+	if _, alreadyMigrated := cache[registerId.Owner]; alreadyMigrated {
+		// yes, return it as is
+		return []ledger.Payload{p}, nil
+	}
+	if !needsMigration(registerId) {
+		return []ledger.Payload{p}, nil
+	}
+
+	// we are going to migrate this
+	cache[registerId.Owner] = struct{}{}
+
+	if em, hasEM := MultipleContractsSpecialMigrations[registerId.Owner]; hasEM {
+		logger.Info().
+			Err(err).
+			Str("address", flow.BytesToAddress([]byte(registerId.Owner)).HexWithPrefix()).
+			Msg("Using exceptional migration for address")
+		return em(p, logger)
+	}
+
+	return migrateValue(p, logger)
+}
+
+func migrateValue(p ledger.Payload, l zerolog.Logger) ([]ledger.Payload, error) {
+
+	// we don't need the the empty code register
+	value := p.Value
+	address := flow.BytesToAddress(p.Key.KeyParts[0].Value)
+	if len(value) == 0 {
+		return []ledger.Payload{}, nil
+	}
+
+	code := string(value)
+	program, err := parser2.ParseProgram(code)
+	if err != nil {
+		l.Error().
+			Err(err).
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("Cannot parse program at address")
+		return nil, err
+	}
+	declarations := program.Declarations
+
+	// find import declarations
```

Comment:
@turbolent does the import part look good to you?

Reply:
Looks good!